File: android/app/src/main/java/com/commuteoptimizer/widget/service/RankingService.kt
package com.commuteoptimizer.widget.service

import com.commuteoptimizer.widget.data.models.CommuteOption
import com.commuteoptimizer.widget.data.models.Weather

/**
 * Ranks commute options with weather-awareness.
 * In bad weather, bike options are demoted below transit options.
 */
object RankingService {

    /**
     * Rank commute options by duration, with weather-aware adjustments.
     * - Sort by total duration (fastest first)
     * - If weather is bad, move first transit option above bike option
     */
    fun rankOptions(options: List<CommuteOption>, weather: Weather): List<CommuteOption> {
        if (options.isEmpty()) return emptyList()

        // Sort by total duration (fastest first)
        val sorted = options.sortedBy { it.durationMinutes }.toMutableList()

        // If bad weather, ensure bike options aren't #1
        if (weather.isBad) {
            val firstBikeIndex = sorted.indexOfFirst { it.type == "bike_to_transit" }
            val firstTransitIndex = sorted.indexOfFirst { it.type == "transit_only" }

            if (firstBikeIndex == 0 && firstTransitIndex > 0) {
                // Move first transit option to #1
                val transit = sorted.removeAt(firstTransitIndex)
                sorted.add(0, transit)
            }
        }

        // Assign ranks
        return sorted.mapIndexed { index, option ->
            option.copy(rank = index + 1)
        }
    }
}

File: ios/CommuteOptimizer/CommuteOptimizer/Services/RankingService.swift
import Foundation

struct RankingService {
    /// Rank options with weather-aware adjustments
    /// - Sort by total duration (fastest first)
    /// - If weather is bad AND bike is #1, swap with first transit option
    static func rankOptions(_ options: [CommuteOption], weather: Weather) -> [CommuteOption] {
        guard !options.isEmpty else { return [] }

        // Sort by duration (fastest first)
        var sorted = options.sorted { $0.durationMinutes < $1.durationMinutes }

        // Weather adjustment: demote bike if weather is bad
        if weather.isBad {
            if let firstBikeIndex = sorted.firstIndex(where: { $0.type == .bikeToTransit }),
               firstBikeIndex == 0,
               let firstTransitIndex = sorted.firstIndex(where: { $0.type == .transitOnly }),
               firstTransitIndex > 0 {
                // Swap: move first transit option to #1
                let transit = sorted.remove(at: firstTransitIndex)
                sorted.insert(transit, at: 0)
            }
        }

        // Assign ranks (1-based)
        return sorted.enumerated().map { index, option in
            var ranked = option
            ranked.rank = index + 1
            return ranked
        }
    }
}
